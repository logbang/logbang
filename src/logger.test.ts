import Logger, { Level } from './logger';

const DateNow = Date.now;

beforeEach(() => {
  Date.now = jest.fn(() => 1514764800000);
});

afterEach(() => {
  Date.now = DateNow;
});

test('basic logging', () => {
  const logger = Logger('testing');
  const spy = jest.spyOn(logger, 'emit');
  logger.log(Level.Info, 'hello world');
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

test('basic logging with payload', () => {
  const logger = Logger('testing');
  const spy = jest.spyOn(logger, 'emit');
  logger.log(Level.Info, 'hello world', { foo: 'bar' });
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

test('basic logging with context and payload', () => {
  const logger = Logger('testing', { foo: 'bar', country: 'uk' });
  const spy = jest.spyOn(logger, 'emit');
  logger.log(Level.Info, 'hello world', { foo: 'baz', cool: true });
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

test('child logger with context and payload', () => {
  const logger = Logger('testing', { foo: 'bar' }).child('testchild', {
    foo: 'qux',
    hello: 'world',
    meh: true,
  });
  const spy = jest.spyOn(logger, 'emit');
  logger.log(Level.Info, 'hello world', { foo: 'baz', cool: true });
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

test('serialized basic errors', () => {
  const logger = Logger('testing');
  const spy = jest.spyOn(logger, 'emit');
  const error = new Error('oops');
  Object.defineProperty(error, 'stack', {
    value: 'deterministic stack\nat line 1\nat line 2',
  });
  logger.log(Level.Info, 'fail whale', { error });
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

test('respects serializable errors', () => {
  const logger = Logger('testing');
  const spy = jest.spyOn(logger, 'emit');
  const error = {
    toJSON() {
      return {
        name: 'serializable error',
        stack: 'nothing to report',
      };
    },
  };
  logger.log(Level.Info, 'fail whale', { error });
  expect(spy.mock.calls[0]).toMatchSnapshot();
});

['error', 'warning', 'info', 'debug'].forEach(level => {
  test(`provides sugar api for ${level}`, () => {
    const logger = Logger('testing');
    const spy = jest.spyOn(logger, 'emit');
    logger[level]('hello world', { foo: 'baz', cool: true });
    expect(spy.mock.calls[0]).toMatchSnapshot();
  });
});
