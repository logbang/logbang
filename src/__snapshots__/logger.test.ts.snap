// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic logging 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {},
    "level": 2,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`basic logging with context and payload 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "country": "uk",
      "foo": "baz",
    },
    "level": 2,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`basic logging with payload 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "foo": "bar",
    },
    "level": 2,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`child logger with context and payload 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "foo": "baz",
      "hello": "world",
      "meh": true,
    },
    "level": 2,
    "logger": "testing:testchild",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`provides sugar api for debug 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "foo": "baz",
    },
    "level": 3,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`provides sugar api for error 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "foo": "baz",
    },
    "level": 0,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`provides sugar api for info 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "foo": "baz",
    },
    "level": 2,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`provides sugar api for warning 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "cool": true,
      "foo": "baz",
    },
    "level": 1,
    "logger": "testing",
    "msg": "hello world",
    "time": 1514764800000,
  },
]
`;

exports[`respects serializable errors 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "error": Object {
        "name": "serializable error",
        "stack": "nothing to report",
      },
    },
    "level": 2,
    "logger": "testing",
    "msg": "fail whale",
    "time": 1514764800000,
  },
]
`;

exports[`serialized basic errors 1`] = `
Array [
  Object {
    "__f": "logbang:1",
    "ctx": Object {
      "error": Object {
        "message": "oops",
        "name": "Error",
        "stack": "deterministic stack
at line 1
at line 2",
      },
    },
    "level": 2,
    "logger": "testing",
    "msg": "fail whale",
    "time": 1514764800000,
  },
]
`;
